src/allmydata/util/deferredutil.py

```diff
@@ -225,16 +226,25 @@ def until(
             break
 
 
-def async_to_deferred(f):
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def async_to_deferred(f: Callable[P, Awaitable[R]]) -> Callable[P, Deferred[R]]:
     """
     Wrap an async function to return a Deferred instead.
 
     Maybe solution to https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3886
     """
 
     @wraps(f)
-    def not_async(*args, **kwargs):
-        return defer.Deferred.fromCoroutine(f(*args, **kwargs))
+    def not_async(*args: P.args, **kwargs: P.kwargs) -> Deferred[R]:
+        # Twisted documents fromCoroutine as accepting either a Generator or a
+        # Coroutine. However, the standard for type annotations of async
+        # functions is to return an Awaitable:
+        # https://github.com/twisted/twisted/issues/11832
+        # So, we ignore the type warning.
```

Comment:
Hmmm.  I'm not sure about this.  What `Deferred.fromCoroutine` actually does with its argument is feed it into the inlineCallbacks machinery ... which actually wants to use `.send` and such.  So maybe `fromCoroutine` isn't actually the right way to do this conversion.  I guess perhaps the reason it has always worked is that the only kind of Awaitable we've ever used it on is the coroutine you get from an `async def`.

Reply:
So ... maybe copying the types from `fromCoroutine` is the right thing to do here?  A more general `awaitable_to_deferred` _is_ possible, but the implementation is different.

Reply:
The example for `fromCoroutine` in the docs takes an async function's result, so...